### PR TITLE
Simplify usages of shardsWithState

### PR DESCRIPTION
**Pull request by idegtiarenko**
This is a followup to https://github.com/elastic/elasticsearch/pull/91991

### CODE REVIEW

**Review by DaveCTurner — APPROVED** [review]
Looks good, assuming CI is happy. Since this is only test changes, can we backport to 8.6 too?

File: server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java [hunk 1]
```diff
@@ -178,7 +178,7 @@ public static <E> ArrayList<E> iterableAsArrayList(Iterable<? extends E> element
             throw new NullPointerException("elements");
         }
         if (elements instanceof Collection) {
-            return new ArrayList<>((Collection) elements);
+            return new ArrayList<>((Collection<E>) elements);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java:181]: Prevent raw usage of parameterized class

File: test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java [hunk 2]
```diff
@@ -12,41 +12,29 @@
...
-            for (RoutingNode routingNode : routingNodes) {
-                shards.addAll(routingNode.shardsWithState(state));
-            }
-        }
-        return shards;
+        return state == ShardRoutingState.UNASSIGNED
+            ? iterableAsArrayList(routingNodes.unassigned())
+            : routingNodes.stream()
+                .flatMap(routingNode -> routingNode.shardsWithState(state).stream())
+                .collect(toCollection(ArrayList::new));
```
Comment (idegtiarenko) [inline test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java:31]: Some tests modify the output of this method (eg shuffle)

File: test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java [hunk 3]
```diff
@@ -12,41 +12,29 @@
...
-                    }
-                }
-            } else {
-                for (RoutingNode routingNode : routingNodes) {
-                    shards.addAll(routingNode.shardsWithState(index, state));
-                }
-            }
-        }
-        return shards;
+    public static List<ShardRouting> shardsWithState(RoutingNodes routingNodes, String index, ShardRoutingState states) {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java:34]: None of the callers of this method need the mutability AFAICT?

File: test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java (hunk 2 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/cluster/routing/RoutingNodesHelper.java:31]: I think I'd rather have those tests copy the result into a new list instead and make this immutable. It's only a couple of spots to change I think.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/DelayedAllocationIT.java [hunk 4]
```diff
@@ -191,7 +190,6 @@ private void indexRandomData() throws Exception {
     private String findNodeWithShard() {
         ClusterState state = client().admin().cluster().prepareState().get().getState();
         List<ShardRouting> startedShards = RoutingNodesHelper.shardsWithState(state.getRoutingNodes(), ShardRoutingState.STARTED);
-        Collections.shuffle(startedShards, random());
-        return state.nodes().get(startedShards.get(0).currentNodeId()).getName();
+        return state.nodes().get(randomFrom(startedShards).currentNodeId()).getName();
```
Comment (idegtiarenko) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/DelayedAllocationIT.java:193]: No need to shuffle entire collection to get a single random element.

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderDecisionTests.java [hunk 5]
```diff
@@ -515,7 +514,6 @@ private void startRandomShards() {
                 allocation.routingNodes(),
                 ShardRoutingState.INITIALIZING
             );
-            initializingShards.sort(Comparator.comparing(ShardRouting::shardId).thenComparing(ShardRouting::primary, Boolean::compare));
             List<ShardRouting> shards = randomSubsetOf(Math.min(randomIntBetween(1, 100), initializingShards.size()), initializingShards);
```
Comment (idegtiarenko) [inline x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderDecisionTests.java:517]: I think this sorting is unnecessary. `randomSubsetOf` shuffles the collection internally.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/DelayedAllocationIT.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/DelayedAllocationIT.java:193]: :+1: even better

File: x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderDecisionTests.java (hunk 5 above)
Comment (DaveCTurner) [reply x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderDecisionTests.java:517]: :+1: even better

**Pull request merged**